lib/Dialect/FIRRTL/Transforms/ResolveTraces.cpp

```diff
@@ -75,49 +75,103 @@ struct ResolveTracesPass : public ResolveTracesBase<ResolveTracesPass> {
+  /// the _next_ index that will be used.
+  unsigned symbolIdx = 0;
 
-    SmallString<64> newTarget("~");
-    newTarget.append(module->getParentOfType<CircuitOp>().getName());
+  /// Map of symbol to symbol index.  This is used to reuse symbol
+  /// substitutions.
+  DenseMap<Attribute, unsigned> symbolMap;
+
+  /// Symbol substitutions for the JSON verbatim op.
+  SmallVector<Attribute> symbols;
+
+  /// Get a symbol index and update symbol datastructures.
+  unsigned getSymbolIndex(Attribute attr) {
+    auto iterator = symbolMap.find(attr);
+    if (iterator != symbolMap.end())
+      return iterator->getSecond();
+
+    auto idx = symbolIdx++;
+    symbolMap.insert({attr, idx});
+    symbols.push_back(attr);
+
+    return idx;
+  }
+
+  /// Convert an annotation path to a string with symbol substitutions.
+  void buildTarget(AnnoPathValue &path, SmallString<64> &newTarget) {
+
+    auto addSymbol = [&](Attribute attr) -> void {
+      newTarget.append("{{");
+      newTarget.append(std::to_string(getSymbolIndex(attr)));
+      newTarget.append("}}");
+    };
+
+    newTarget.append("~");
+    newTarget.append(
+        path.ref.getModule()->getParentOfType<CircuitOp>().getName());
     newTarget.append("|");
 
-    if (auto nla = anno.getMember<FlatSymbolRefAttr>("circt.nonlocal")) {
-      hw::HierPathOp path = nlaTable->getNLA(nla.getAttr());
-      for (auto part : path.getNamepath().getValue().drop_back()) {
-        auto inst = cast<hw::InnerRefAttr>(part);
-        newTarget.append(inst.getModule());
-        newTarget.append("/");
-        newTarget.append(inst.getName());
-        newTarget.append(":");
-      }
+    if (path.isLocal()) {
+      addSymbol(
+          FlatSymbolRefAttr::get(path.ref.getModule().getModuleNameAttr()));
+    } else {
+      addSymbol(FlatSymbolRefAttr::get(path.instances.front()
+                                           ->getParentOfType<FModuleLike>()
+                                           .getModuleNameAttr()));
+    }
+
+    for (auto inst : path.instances) {
+      newTarget.append("/");
+      addSymbol(hw::InnerRefAttr::get(
+          inst->getParentOfType<FModuleLike>().getModuleNameAttr(),
+          inst.getInnerSymAttr().getSymName()));
+      newTarget.append(":");
+      addSymbol(inst.getModuleNameAttr());
     }
-    newTarget.append(module.getModuleName());
-    newTarget.append(">");
 
-    newTarget.append(name);
+    // If this targets a module or an instance, then we're done.  There is no
+    // "reference" part of the FIRRTL target.
+    if (path.ref.isa<OpAnnoTarget>() &&
+        path.isOpOfType<FModuleOp, FExtModuleOp, InstanceOp>())
+      return;
 
-    // Add information if this is an aggregate.
-    auto targetFieldID = anno.getFieldID();
+    newTarget.append(">");
+    auto innerSym =
+        TypeSwitch<AnnoTarget, hw::InnerSymAttr>(path.ref)
+            .Case<PortAnnoTarget>([&](PortAnnoTarget portTarget) {
```

Comment:
Is it known/required these already have inner symbols on them? (I'm guessing "yes")?

(and is this why not using, e.g., `AnnoTarget::getInnerSym`?)

Reply:
That would be better!